### DISCUSSION TITLE
Sending fullTopic in rule

### DISCUSSION THREAD

**Discussion by guydvir2** [arendst/Tasmota#11236] [Q&A] [answered]
Hi,
I'm trying to figure out how to send the value of FullTopic as defined in MQTT section, when posting a MQTT msg :+1: 

For example:
`rule1 ON POWER1#state DO backlog POWER2 %value%; publish myHome/Messages [%timestamp%] %fullTopic% %value%  ENDON`

**Comment by barbudor** [comment]
If what you need is just the result of the command `Fulltopic` for logging you can set a boot time rule such as:
`on system#boot do backlog fulltopic endon on fulltopic#data do var1 %value% endon`
This will capture the fulltopic in a variable that you can re-use in your logging rule.

**Reply by guydvir2** [reply]
Can you please explain what is the first part of the rule `on system#boot do backlog fulltopic endon` ?
Second part is clear.

BTW
creating a rule - `on system#boot do backlog fulltopic endon on fulltopic#data do var1 %value% endon`, and testing it after reboot using console `publish myHome/Message var1`, yielded `var1` at the specific topic and not what I meant :/

**Reply by barbudor** [reply]
'on system#boot' is a boot time trigger
'backlog fulltopic' execute 'fulltopic' command and force the resulting mqtt message to go through the rule parser
'on fulltopic#data' triggers on the message resulting from the above 'fulltopic' command
'var1 %value%' store the result of 'fulltopic' command into var1

You should use '%var1%' in a rule (and not in the command line) to use the fulltopic because %% are replaced only in Rules execution, not in command line

**Reply by guydvir2** [reply]
regarding 2nd line: `backlog fulltopic` - why a use of `backlog` if it is not a sequence of commands (no semicolon) ?
reset of the rule was understandable.

**Reply by sfromis** [reply]
Because of "force the resulting mqtt message to go through the rule parser". The thing is that when a command is invoked from a rule, the output does not always gets passed to the JSON parser needed for the triggers to invoke other rules. This is mitigated by the extra nesting level from using Backlog.

**Comment by sfromis** [comment, marked as answer]
The "problem" with fulltopic is that it is not a static value, but depending on circumstances the `%prefix%/%topic%` gets to be `stat/%topic%` or `tele/%topic%` (or the other order if you've selected that). Thus you do not have `%prefix%` available in rules, but in most cases, you can just set the prefix yourselves, using something like `publish stat/%topic%/mypower %value%` and so on.

**Reply by barbudor** [reply]
I originally made long post about that but finally my understanding is that he just want to log the value of fulltopic and not use that as a publication topic

**Reply by sfromis** [reply]
I was assuming the wish was not to have a literal string like `%prefix%/tasmotaname/` included, makes no difference if it is in topic or payload. That unexpanded `%prefix%` is probably not useful in the payload. Simpler to just use `%topic%` in the rule, which will expand. Note that to actually capture the output of the `FullTopic` command, you'd need to use 
`on fulltopic#data do var1 %value% endon`
Yeah, pesky detail with that `#data` but that's how it works :-)

**Reply by guydvir2** [reply]
Thank your for your replies. `FullTopic` describes iot device as: `myHome/internalLights/Corridor`. I have a topic that all actions are forwarded to which call `myHome/Messages` which every device that makes an action, post to that topic as: 
`[clock][fullTopic] <action>`

**Reply by sfromis** [reply]
Generally, the intention is that you use Topic to identify your device, instead of putting it all into FullTopic. This works best with various Tasmota features which allows [extra topics](https://tasmota.github.io/docs/MQTT/#mqtt-topic-definition) for specific functions, when you include %topic% in FullTopic. Basically, your scheme cuts out a lot of Tasmota features.

**Reply by guydvir2** [reply]
my fullTopic is: `myHome/Lights/%topic%/%prefix% - what is lost the way I use?

**Reply by sfromis** [reply]
Then your fulltopic is not what you wrote, `myHome/internalLights/Corridor`. If you instead have `myHome/Lights/%topic%/%prefix%` and via the command and rule gets it into a variable, that string would include literally `%prefix%`, not expanded.
